pkg/lib/oauth/handler/resolve.go

```diff
@@ -101,3 +102,64 @@ func validateRedirectURI(
 
 	return nil
 }
+
+func parseAuthzRedirectURI(
+	client *config.OAuthClientConfig,
+	httpProto httputil.HTTPProto,
+	httpOrigin httputil.HTTPOrigin,
+	domainWhitelist []string,
+	r protocol.AuthorizationRequest,
+) (*url.URL, protocol.ErrorResponse) {
+	if r.SettingsAction() == "" {
+		return parseRedirectURI(client, httpProto, httpOrigin, domainWhitelist, r)
+	}
+
+	redirectURI, err := url.Parse(r.RedirectURI())
+	if err != nil {
+		return nil, protocol.NewErrorResponse("invalid_request", "invalid redirect URI")
+	}
+
+	err = validatePostSettingsRedirectURI(client, httpProto, httpOrigin, domainWhitelist, redirectURI)
+	if err != nil {
+		return nil, protocol.NewErrorResponse("invalid_request", err.Error())
+	}
+
+	switch r.SettingsAction() {
+	case "change_password":
+		settingsRedirectURI := &url.URL{Path: "/settings/change_password"}
```

Comment:
We should follow the same logic used in UIURLBuilder to build the URI.

1. UIURLBuilder.Build currently build authentication URL. Let's rename it to BuildAuthenticationURL
2. Add UIURLBuilder.BuildSettingsURL for building settings URL.